proposals/p3762.md

```diff
@@ -0,0 +1,828 @@
+make code clearer by helping developers catch redundant, and possibly incorrect,
+code.
+
+### `extern` naming
+
+Beyond `extern`, we also considered `external`. `extern` implies external
+linkage in C++. We're choosing `extern` mainly for the small consistency with
+C++.
+
+### Default `extern` to private
+
+The `extern` keyword could have an access control implication equivalent to
+`private`. Then `extern` would need explicit work to export the symbol. The
+`export` keyword was proposed for this purpose, with the idea that
+`export import` syntax might also be provided to re-export all symbols of an
+imported library.
+
+1.  `extern` and `export` semantic consistency
+
+    This would mean that `extern` declarations have different access control
+    than other declarations, which is a different visibility model to
+    understand, and may also not be intuitive from the "external" meaning. The
+    use of `export` matches C++'s `export` keyword, which may also imply to
+    developers that other semantics match C++, such as `export` being necessary
+    for all declared names.
+
+2.  Interaction with additional access control features
+
+    We'll probably also want more granular access control than just public and
+    private for API names. Adding package-private access modifier seems useful
+    (for example, Java provides this as `package`): in C++, this is sometimes
+    achieved through an "internal" or "details" namespace. If Carbon only
+    supports library-private symbols, that still addresses some of these
+    use-cases, but will sometimes require private implementation details to
+    exist in a single `api` file in order to get language-supported visibility
+    restrictions. In some cases this will result in an unwieldy amount of code
+    for a single file.
+
+    For example of how package-private visibility might be used, gtest has
+    [an internal header directory](https://github.com/google/googletest/blob/main/googletest/include/gtest/internal)
+    that contains thousands of lines of code. If this needed to migrate to `api`
+    files in Carbon, it would be ideal if users could not access the names by
+    importing an internal library.
+
+    For example of how this would interact:
+
+    | Default visibility             | Public (proposed)         | Private (alternative)            |
+    | ------------------------------ | ------------------------- | -------------------------------- |
+    | Public                         | `extern class C;`         | `export extern class C;`         |
+    | Library-private                | `private extern class C;` | `extern class C;`                |
+    | Package-private (hypothetical) | `package extern class C;` | `export package extern class C;` |
+
+3.  Risks of a public extern
+
+    When making an `extern fn`, it is callable. This creates a risk of a
+    function being declared as `extern` for intern use, but accidentally
```

Comment:
```suggestion
    function being declared as `extern` for internal use, but accidentally
```

I think restrictions on which functions interns can call are best enforced by code review rather than language rules ;-)